### ISSUE TITLE
Bug: [prefer-optional-chain] Invalid fix when comparing against `null`

### ISSUE THREAD

**Issue by szuend** [typescript-eslint/typescript-eslint#12204]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=6.0.2&fileType=.tsx&code=FASwdgLgpgTgZgQwMZQAQDED2nUG9iqqYCu0MAtpgM4RaYAUAlAFwbYA%2BYxANt8AL7BgcYmCQQQmMKiggIAC1h0A8jAByPbk1Z1U7VF154CqGFAjEY0gLIIFAOhgIwAE0zkmqADyoADPYBWVAB%2BPBIySho6Vk8AXgA%2BA01%2BVFZDbgBuASEkKRpUOGxUWJk5RRgVdU0mLJA4VHoAQkKcdn0W%2B3DYSNpsT0bYkpbGY0IAejHUABEcKndzeXAAc3tsoA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WloHsalfkwCG8WmQAWo5uii9o-aJHBgAviHVA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```TypeScript
interface Foo {
  outermostFoo(): Foo|null
}

function eitherFooOrNull(): Foo | null {
  return Math.random() < 0.5 ? {outermostFoo: () => null} : null;
}

const foo = eitherFooOrNull();
if (!foo || foo.outermostFoo() !== foo) {
  // Do something.
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/prefer-optional-chain": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

No error/fix happening.

### Actual Result

The if is re-written

```ts
const foo = eitherFooOrNull();
if (foo?.outermostFoo() !== foo) {
  // Do something.
}
```

This is a true semantic change, because optional chaining coerces the `null` into an `undefined`, which means in the case where `foo` is `null`, the if changes to `undefined !== null`. We will no longer take the if branch.

### Additional Info

Broken in the upgrade from 8.47.0 to 8.52.0. Upgrading typescript-eslint in the Chrome DevTools code base causes actual test failures due to this rewrite and we had to disable the rule.

**Comment by szuend** [comment]
I guess duplicate of issue #12200 .

Gonna be honest here, we auto apply fixes as part of our presubmit script, which means this rule is now not worth enabling given the high degree of false positives we saw.

**Comment by JoshuaKGoldberg** [comment]
Interesting. For context, rule reports can include either _fixes_ and _suggestions_, and those two are not the same:

* **Fixes** are what are applied by running ESLint with `--fix`. They are supposed to always be safe (not change code behavior in a potentially buggy way). A rule report may include 0 or 1 of these.
* **Suggestions** are also code changes, but they don't have to be safe. They are not applied even with `--fix`. A rule report may include any number of these.

See also: https://eslint.org/docs/latest/use/core-concepts/glossary#fix, https://eslint.org/docs/latest/use/core-concepts/glossary#suggestion

This issue shows an unsafe _fix_ in a rule report. I could consider that a bug.

Unless something has changed since it was filed, the playground from #12200 also shows an unsafe fix - _not_ a suggestion. So I'd consider both of these issues to be valid bugs to be fixed. @bradzacher am I missing something?

**Comment by kirkwaiblinger** [comment]
Might be a duplicate/special case of https://github.com/typescript-eslint/typescript-eslint/issues/11840 though?

**Comment by bradzacher** [comment]
If I missed something and saw a suggestion instead of an autofix then yeah both are bugs.

**Comment by bradzacher** [comment]
But looks like a duplicate of #11840

**Comment by szuend** [comment]
Just to be clear, we are getting **invalid fixes**, and not wrong suggestions. Having wrong suggestions would be fine, but the invalid fix means that we had to disable the rule.
